File: frontend/src/components/Settings/presets.ts
import type { CreateAgent } from '../../api/client'

/**
 * 预设 Agent 模板（静态、低频变更）。
 *
 * 数据来源：复用 `/home/pax/coding/research/obsidian-agent-client`
 * 的 ACP 客户端预设，包含两部分：
 *   1. `src/plugin.ts` 的 `DEFAULT_SETTINGS` 内置客户端
 *      （Claude / Codex / Gemini）；
 *   2. `docs/agent-setup/custom-agents.md` 额外推荐的自定义 agent 示例
 *      （OpenCode / Qwen Code / Kiro）。
 * 仅映射 OmniTerm 支持的字段（display_name / command / args / env）。
 * 原项目带有 `apiKeySecretId`（指向 Obsidian 密钥存储），OmniTerm 无该
 * 体系，凭据由 agent 进程自身负责（见 api/client.ts `Agent` 注释），故不映射。
 *
 * 增/改预设改本文件即可；说明文案（参考说明）走两个
 * `frontend/src/locales/{en,zh}/translation.json` 的 `settings.agents.preset.*` key。
 */

export interface AgentPreset extends CreateAgent {
  /** 面板展示用的简短标签（i18n key） */
  labelKey: string
  /** 一键填入后的提示文案（i18n key） */
  hintKey: string
}

export const AGENT_PRESETS: AgentPreset[] = [
  {
    labelKey: 'settings.agents.preset.claude.label',
    hintKey: 'settings.agents.preset.claude.hint',
    display_name: 'Claude Code',
    command: 'claude-agent-acp',
    args: [],
    env: [],
  },
  {
    labelKey: 'settings.agents.preset.codex.label',
    hintKey: 'settings.agents.preset.codex.hint',
    display_name: 'Codex',
    command: 'codex-acp',
    args: [],
    env: [],
  },
  {
    labelKey: 'settings.agents.preset.gemini.label',
    hintKey: 'settings.agents.preset.gemini.hint',
    display_name: 'Gemini CLI',
    command: 'gemini',
    args: ['--experimental-acp'],
    env: [],
  },
  {
    labelKey: 'settings.agents.preset.opencode.label',
    hintKey: 'settings.agents.preset.opencode.hint',
    display_name: 'OpenCode',
    command: 'opencode',
    args: ['acp'],
    env: [],
  },
  {
    labelKey: 'settings.agents.preset.qwen.label',
    hintKey: 'settings.agents.preset.qwen.hint',
    display_name: 'Qwen Code',
    command: 'qwen',
    args: ['--experimental-acp'],
    env: [],
  },
  {
    labelKey: 'settings.agents.preset.kiro.label',
    hintKey: 'settings.agents.preset.kiro.hint',
    display_name: 'Kiro',
    command: 'kiro-cli',
    args: ['acp'],
    env: [],
  },
]

File: src/presets.rs
/// Built-in agent presets for first-run experience.
///
/// On startup, OmniTerm checks whether each preset's `command` is findable on
/// PATH (via `which` on Unix / `where` on Windows). Only commands that are
/// actually installed are inserted into the `agents` table (INSERT OR IGNORE
/// keeps subsequent re-runs idempotent and preserves user edits).
///
/// The same preset data is mirrored in `frontend/src/components/Settings/presets.ts`
/// so Settings can render "fill-in" buttons for users who deleted a preset and
/// want to restore it (or added a previously-missing CLI later).
use sqlx::SqlitePool;
use std::process::Command;

struct AgentPreset {
    id: &'static str,
    display_name: &'static str,
    command: &'static str,
    args_json: &'static str,
}

const BUILTIN_PRESETS: &[AgentPreset] = &[
    AgentPreset {
        id: "preset-claude",
        display_name: "Claude Code",
        command: "claude-agent-acp",
        args_json: "[]",
    },
    AgentPreset {
        id: "preset-codex",
        display_name: "Codex",
        command: "codex-acp",
        args_json: "[]",
    },
    AgentPreset {
        id: "preset-gemini",
        display_name: "Gemini CLI",
        command: "gemini",
        args_json: r#"["--experimental-acp"]"#,
    },
    AgentPreset {
        id: "preset-opencode",
        display_name: "OpenCode",
        command: "opencode",
        args_json: r#"["acp"]"#,
    },
    AgentPreset {
        id: "preset-qwen",
        display_name: "Qwen Code",
        command: "qwen",
        args_json: r#"["--experimental-acp"]"#,
    },
    AgentPreset {
        id: "preset-kiro",
        display_name: "Kiro",
        command: "kiro-cli",
        args_json: r#"["acp"]"#,
    },
    AgentPreset {
        id: "preset-qoder",
        display_name: "qoder",
        command: "qodercli",
        args_json: r#"["--acp"]"#,
    },
    AgentPreset {
        id: "preset-qoder-cn",
        display_name: "qoder cn",
        command: "qoderclicn",
        args_json: r#"["--acp"]"#,
    },
    AgentPreset {
        id: "preset-oh-my-pi",
        display_name: "oh-my-pi",
        command: "omp",
        args_json: r#"["acp"]"#,
    },
];

/// Test whether `name` is findable on PATH.
fn command_exists(name: &str) -> bool {
    #[cfg(windows)]
    let (checker, arg) = ("where", name);
    #[cfg(not(windows))]
    let (checker, arg) = ("which", name);

    Command::new(checker)
        .arg(arg)
        .stdout(std::process::Stdio::null())
        .stderr(std::process::Stdio::null())
        .status()
        .map(|s| s.success())
        .unwrap_or(false)
}

/// Seed built-in agent presets for commands that are actually installed.
///
/// Uses `INSERT OR IGNORE` so re-running on the same DB is a no-op and
/// user-edited presets are never overwritten.
pub async fn seed_builtin_presets(db: &SqlitePool) {
    for preset in BUILTIN_PRESETS {
        if !command_exists(preset.command) {
            continue;
        }
        let now = chrono::Utc::now().to_rfc3339();
        let _ = sqlx::query(
            "INSERT OR IGNORE INTO agents (id, display_name, command, args, env, created_at, updated_at) \
             VALUES (?, ?, ?, ?, '[]', ?, ?)",
        )
        .bind(preset.id)
        .bind(preset.display_name)
        .bind(preset.command)
        .bind(preset.args_json)
        .bind(&now)
        .bind(&now)
        .execute(db)
        .await;
    }
}
